File: python_version/validation_helper.py
import re
from datetime import datetime


class ValidationHelper:
    def is_null_or_white_space(self, value):
        return value is None or str(value).strip() == ""

    def is_valid_email(self, email):
        if self.is_null_or_white_space(email):
            return False

        normalized = email.strip()
        pattern = r"^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}$"
        return bool(re.fullmatch(pattern, normalized)) and normalized == email

    def read_required_string(self, prompt):
        while True:
            value = input(prompt)
            if not self.is_null_or_white_space(value):
                return value.strip()
            print("Input cannot be empty.")

    def read_employee_id(self, prompt):
        while True:
            value = self.read_required_string(prompt)
            if len(value) >= 1:
                return value
            print("Employee ID cannot be empty.")

    def read_valid_date(self, prompt):
        while True:
            value = input(prompt)
            try:
                return datetime.strptime(value.strip(), "%Y-%m-%d").date()
            except ValueError:
                print("Please enter a valid date.")

File: EmployeeManagement/Utilities/ValidationHelper.cs
namespace EmployeeManagement.Utilities;

public class ValidationHelper
{
    public bool IsNullOrWhiteSpace(string? value) => string.IsNullOrWhiteSpace(value);

    public bool IsValidEmail(string? email)
    {
        if (string.IsNullOrWhiteSpace(email))
        {
            return false;
        }

        try
        {
            var addr = new System.Net.Mail.MailAddress(email);
            return addr.Address == email.Trim();
        }
        catch
        {
            return false;
        }
    }

    public string ReadRequiredString(string prompt)
    {
        while (true)
        {
            Console.Write(prompt);
            var input = Console.ReadLine();

            if (!string.IsNullOrWhiteSpace(input))
            {
                return input.Trim();
            }

            Console.WriteLine("Input cannot be empty.");
        }
    }

    public string ReadEmployeeId(string prompt)
    {
        while (true)
        {
            var input = ReadRequiredString(prompt);
            if (input.Length >= 1)
            {
                return input;
            }

            Console.WriteLine("Employee ID cannot be empty.");
        }
    }

    public DateTime ReadValidDate(string prompt)
    {
        while (true)
        {
            Console.Write(prompt);
            var input = Console.ReadLine();

            if (DateTime.TryParse(input, out var parsedDate))
            {
                return parsedDate;
            }

            Console.WriteLine("Please enter a valid date.");
        }
    }
}
